### PR TITLE
(LTH-140) Download file from a given URL

### PR DESCRIPTION
**Pull request by ekinanp**
Leatherman.curl has been extended to now download a file from a
provided URL and write the downloaded contents to a user-specified
file path. Some refactoring was also done in Leatherman.curl to
take out common error-checking patterns.

NOTE: I am going to start writing tests for this PR. This is just to show the main logic and the refactoring.

### CODE REVIEW

File: curl/src/client.cc [hunk 1]
```diff
@@ -296,63 +314,54 @@ namespace leatherman { namespace curl {
...
-                throw http_request_exception(ctx.req, curl_easy_strerror(result));
-            }
+        if (_ca_cert == "") {
+            return;
         }
+        
+        curl_easy_setopt_maybe<const char*>(ctx, CURLOPT_CAINFO, _ca_cert.c_str());
+    }
+
+    void client::set_ca_info(std::string const& url, std::string const& file_path){
```
Comment (MikaelSmith) [inline curl/src/client.cc]: What is this doing? It makes no sense to me.

File: curl/inc/leatherman/curl/client.hpp [hunk 2]
```diff
@@ -213,13 +262,41 @@ namespace leatherman { namespace curl {
         LEATHERMAN_CURL_NO_EXPORT void set_timeouts(context& ctx);
         LEATHERMAN_CURL_NO_EXPORT void set_write_callbacks(context& ctx);
         LEATHERMAN_CURL_NO_EXPORT void set_client_info(context &ctx);
+        LEATHERMAN_CURL_NO_EXPORT void set_client_info(std::string const& url, std::string const& file_path);
         LEATHERMAN_CURL_NO_EXPORT void set_ca_info(context& ctx);
+        LEATHERMAN_CURL_NO_EXPORT void set_ca_info(std::string const& url, std::string const& file_path);
         LEATHERMAN_CURL_NO_EXPORT void set_client_protocols(context& ctx);
 
+        template <typename ParamType> 
+        LEATHERMAN_CURL_NO_EXPORT void curl_easy_setopt_maybe(
```
Comment (MikaelSmith) [inline curl/inc/leatherman/curl/client.hpp]: Refactoring around this is better done as a separate commit.

File: curl/src/client.cc (hunk 1 above)
Comment (ekinanp) [inline curl/src/client.cc]: I can one line it, was a wrapper to curl_easy_setopt_maybe. I wanted to preserve the existing pattern in perform where they set_ca_info, then set_client_info, etc.

File: curl/inc/leatherman/curl/client.hpp (hunk 2 above)
Comment (ekinanp) [inline curl/inc/leatherman/curl/client.hpp]: Do you want me to make it a separate commit? I can make the refactoring for the main code base (before the download_file) as one commit, then add the download_file as another commit.

File: curl/src/client.cc (hunk 1 above)
Comment (MikaelSmith) [inline curl/src/client.cc]: Oh, it's a helper for the exception that will be thrown. That's somewhat confusing. I'd be ok with set_ca_info just throwing one type of exception.

File: curl/inc/leatherman/curl/client.hpp (hunk 2 above)
Comment (MikaelSmith) [inline curl/inc/leatherman/curl/client.hpp]: Yes please.

File: curl/inc/leatherman/curl/client.hpp [hunk 3]
```diff
@@ -108,6 +108,48 @@ namespace leatherman { namespace curl {
...
+     */
+    struct LEATHERMAN_CURL_EXPORT http_file_download_exception : http_exception
+    {
+        /**
+         * Constructs an http_file_download_exception.
+         * @param url The url that caused the exception.
+         * @param file_path The file that was meant to be downloaded
+         * @param message The exception message.
+         */
+        http_file_download_exception(std::string const &url, std::string const &file_path, std::string const &message) :
```
Comment (MikaelSmith) [inline curl/inc/leatherman/curl/client.hpp]: The standard pattern for construction is to pass by value, then move into the object. You're mixing them by passing by reference, which is causing an extra copy.

File: curl/src/client.cc (hunk 1 above)
Comment (MikaelSmith) [inline curl/src/client.cc]: I think this is mostly a case of non-obvious behavior. I'd make the name clearer about what it's doing.

File: curl/src/client.cc [hunk 4]
```diff
@@ -133,14 +138,8 @@ namespace leatherman { namespace curl {
...
         // Set common options
-        auto result = curl_easy_setopt(_handle, CURLOPT_NOPROGRESS, 1);
-        if (result != CURLE_OK) {
-            throw http_request_exception(req, curl_easy_strerror(result));
-        }
-        result = curl_easy_setopt(_handle, CURLOPT_FOLLOWLOCATION, 1);
-        if (result != CURLE_OK) {
-            throw http_request_exception(req, curl_easy_strerror(result));
-        }
+        curl_easy_setopt_maybe<long>(ctx, CURLOPT_NOPROGRESS, 1);
```
Comment (MikaelSmith) [inline curl/src/client.cc]: The template parameter should be inferred from the arguments. No need to specify the `<long>`. Should be able to avoid defining all the callback signatures as well.

File: locales/leatherman.pot [hunk 5]
```diff
@@ -18,32 +18,51 @@ msgstr ""
 "Content-Transfer-Encoding: 8bit\n"
 
 #: curl/src/client.cc
+#: curl/src/client_with_download.cc
```
Comment (MikaelSmith) [inline locales/leatherman.pot]: Picked up some extra entries. Looks better though.

Comment (ekinanp) [inline locales/leatherman.pot]: Yeah, those should go away once I get the download stuff up there.

**Comment by MikaelSmith** [comment]
Couple issues with cpplint, and something funny's going on with the Windows build. I dig more tomorrow.

**Comment by ekinanp** [comment]
Ok put up an updated PR with tests. I'm not sure if I should test for all the failure cases of curl_easy_setopt like the existing curl tests do, but I can if that is necessary. I have also tested the download code by making a request to the orchestrator on my master vm and it worked -- the SSL was set up correctly, and once I added the X-Authentication token, the response was downloaded as a file. I'm thinking it might not be a bad idea to add the token as a field to the client class instead of as a parameter to the download method, since it seems like something that'd belong w/ it.

**Comment by MikaelSmith** [comment]
Ah, two Windows issues: we need to add the dependency on Boost.Filesystem to curl/CMakeLists.txt. Append `filesystem system` to the 1st line.

Also tests fail on Windows. Looks like you might be re-using a temporary file.

**Comment by MikaelSmith** [comment]
https://github.com/puppetlabs/orchestrator-client/blob/hoyt/lib/src/client.cc#L122 is how orchestrator-client adds the X-Auth token. I'm surprised that's needed for the static file download though.

File: curl/src/client.cc [hunk 6]
```diff
@@ -173,6 +171,65 @@ namespace leatherman { namespace curl {
...
+            curl_easy_setopt_maybe(url, file_path, CURLOPT_WRITEDATA, fp);
+
+            // TODO: How long should connection time be? Is 5 minutes too much?
+            curl_easy_setopt_maybe(url, file_path, CURLOPT_CONNECTTIMEOUT, 300);
+            set_ca_info(url, file_path);
+            set_client_info(url, file_path);
+
+            // Authentication header
+            curl_list x_auth_header;
+            x_auth_header.append("X-Authentication: " + token);
```
Comment (MikaelSmith) [inline curl/src/client.cc]: This shouldn't be necessary. It can also be handled externally with add_headers.

File: curl/src/client.cc [hunk 7]
```diff
@@ -173,6 +171,65 @@ namespace leatherman { namespace curl {
...
+            auto fp = boost::nowide::fopen(temp_path.string().c_str(), "wb");
+            if (!fp) {
+                throw http_file_download_exception(url, file_path, _("Failed to open temporary file for writing"));
+            }
+
+            curl_easy_setopt_maybe(url, file_path, CURLOPT_WRITEFUNCTION, write_file);
+            curl_easy_setopt_maybe(url, file_path, CURLOPT_WRITEDATA, fp);
+
+            // TODO: How long should connection time be? Is 5 minutes too much?
+            curl_easy_setopt_maybe(url, file_path, CURLOPT_CONNECTTIMEOUT, 300);
```
Comment (MikaelSmith) [inline curl/src/client.cc]: This is handled by the caller via set_timeouts.

File: curl/src/client.cc [hunk 8]
```diff
@@ -173,6 +171,65 @@ namespace leatherman { namespace curl {
         return res;
     }
 
+    void client::download_file(std::string const& url, std::string const& file_path, std::string const& token)
```
Comment (MikaelSmith) [inline curl/src/client.cc]: I'd have the signature take a request object, and a file_path.

**Comment by ekinanp** [comment]
If I have a printf statement in there, that's to help me see where Travis fails. For some reason, the Windows part in appveyor is passing, but Travis release is getting a segfault at the fopen failure test, so I've been trying to figure out why. Anyways TL;DR, the printf statement will be removed.

File: curl/src/client.cc [hunk 9]
```diff
@@ -6,12 +6,16 @@
 #include <leatherman/locale/locale.hpp>
 #include <boost/utility/string_ref.hpp>
 #include <boost/algorithm/string.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/nowide/cstdio.hpp>
 #include <sstream>
+#include <stdio.h>
```
Comment (MikaelSmith) [inline curl/src/client.cc]: Is this for `fclose`? I'd prefer including via `cstdio` for C++.

File: curl/src/client.cc [hunk 10]
```diff
@@ -173,6 +171,79 @@ namespace leatherman { namespace curl {
...
+            if (result == CURLE_OK) {
+                  printf("Error was before download.\n");
+              throw http_file_download_exception(req, file_path, e.what());
+            }
+                 printf("Error was in temp file removal\n"); 
+
+            throw http_file_download_exception(
+                req,
+                file_path,
+                errbuf + _(" and failed to remove temporary file"));
```
Comment (MikaelSmith) [inline curl/src/client.cc]: This should use string substitution if we want it to translate usefully. As in `_("{1} and failed to remove temporary file", errbuf)`.

File: curl/src/client.cc [hunk 11]
```diff
@@ -173,6 +171,79 @@ namespace leatherman { namespace curl {
...
+
+            LOG_DEBUG("download completed, now writing result to file");
+            fs::rename(temp_path, file_path);
+        } catch (http_file_download_exception& e) {
+                 printf("Threw a file_download exception, re-throwing it. Message is: %s\n", e.what());
+            // so that the code in the lower catch block is not hit
+            throw e;
+        } catch (http_request_exception& e) {
+                 printf("Threw a request exception, converting to download. Message is: %s\n", e.what());
+            fclose(fp);
```
Comment (MikaelSmith) [inline curl/src/client.cc]: This should also ensure the file is removed.

Comment (ekinanp) [inline curl/src/client.cc]: You're right, making the fix now.

File: .travis.yml [hunk 12]
```diff
@@ -1,4 +1,5 @@
 language: cpp
+dist: precise
```
Comment (MikaelSmith) [inline .travis.yml]: Seems like this wasn't needed to fix the test.

File: curl/inc/leatherman/curl/client.hpp [hunk 13]
```diff
@@ -153,6 +206,13 @@ namespace leatherman { namespace curl {
         response put(request const& req);
 
         /**
+         * Downloads the file from the specified url.
+         * @param req The HTTP request to perform 
+         * @param file_path The file that the downloaded contents will be written to. 
+         */
+        void download_file(request const& req, std::string const& file_path);
```
Comment (MikaelSmith) [inline curl/inc/leatherman/curl/client.hpp]: Since we're writing a file, we should probably allow specifying the permissions as well. I can handle that in https://github.com/puppetlabs/leatherman/pull/249 as well if you'd prefer.

File: curl/src/client.cc [hunk 14]
```diff
@@ -173,6 +170,85 @@ namespace leatherman { namespace curl {
         return res;
     }
 
+    void client::download_file(request const& req, std::string const& file_path)
+    {
+                   printf("Entered download_file\n");
```
Comment (MikaelSmith) [inline curl/src/client.cc]: Lots of printf's to remove.

File: .travis.yml (hunk 12 above)
Comment (MikaelSmith) [inline .travis.yml]: Turns out I was wrong, we'd just been getting Precise builds until now.

File: curl/inc/leatherman/curl/client.hpp [hunk 15]
```diff
@@ -216,10 +276,25 @@ namespace leatherman { namespace curl {
         LEATHERMAN_CURL_NO_EXPORT void set_ca_info(context& ctx);
         LEATHERMAN_CURL_NO_EXPORT void set_client_protocols(context& ctx);
 
+        LEATHERMAN_CURL_NO_EXPORT void cleanup_temp_file(boost::filesystem::path const& temp_path, FILE* tfp, request const& req, std::string const& file_path, std::string const& reason);
```
Comment (MikaelSmith) [inline curl/inc/leatherman/curl/client.hpp]: I'm going to clean it up separately, but this could be done as a forward-declaration rather than including the header file.
